vowpalwabbit/queue.h

```diff
@@ -56,29 +55,33 @@ class ptr_queue
     {
       is_not_full.wait(lock);
     }
-
     object_queue.push(item);
 
-    lock.unlock();
     is_not_empty.notify_all();
   }
 
   void set_done()
   {
-    done = true;
-
+    {
+      std::unique_lock<std::mutex> lock(mut);
+      done = true;
+    }
```

Comment:
Interesting.... Did this cause real bugs?

In essence, I'm unsure what counts as an atomic operation on a modern computer.

Reply:
There weren't any bugs regarding this. These were issues I found while trying to figure out the race condition I was seeing (the one that was in neural net)

In general I don't think anything thats not explicitly called out as atomic can be relied on to be such. In this case, we might be able to get away with converting the done boolean to a `std::atomic<bool>`, but i'd have to do some more analysis on the consequences of that (especially on the read performance since this lock should only be hit once)

Reply:
Where is the list of things explicitly called out as atomic?

Reply:
std::atomic is a templated type. I believe all the primitive types have specializations while objects use a generic implementation.

According to https://en.cppreference.com/w/cpp/atomic/atomic there are specializations for pointers (including weak_ptr and shared_ptr), booleans, all integral types, and float types